examples/prom/nonlinear_elasticity_global_rom.cpp

```diff
@@ -1691,16 +1912,812 @@ void RomOperator::Mult_FullOrder(const Vector& vx, Vector& dvx_dt) const
+        CalcInverse(Trans.Jacobian(), Jrt);
+        fe.CalcDShape(ip, DSh);
+        Mult(DSh, Jrt, DS);
+        MultAtB(PMatI, DS, Jpt);
+        model->EvalP(Jpt, P);
+        P *= t; // NB: Not by ip.weight
+        AddMultABt(DS, P, PMatO);
+
+        r[i] = 0.0;
+
+        // Calculate r[i] = ve_j^T * elvect
+        for (int k = 0; k < elvect.Size(); k++)
+        {
+            r[i] += ve_j[k] * elvect[k];
+        }
+    }
+}
+
+void SolveNNLS(const int rank, const double nnls_tol, const int maxNNLSnnz,
+               CAROM::Vector const &w, CAROM::Matrix &Gt,
+               CAROM::Vector &sol)
+{
+    CAROM::NNLSSolver nnls(nnls_tol, 0, maxNNLSnnz, 2);
+
+    CAROM::Vector rhs_ub(Gt.numColumns(), false);
+    // G.mult(w, rhs_ub);  // rhs = Gw
+    // rhs = Gw. Note that by using Gt and multTranspose, we do parallel communication.
+    Gt.transposeMult(w, rhs_ub);
+
+    CAROM::Vector rhs_lb(rhs_ub);
+    CAROM::Vector rhs_Gw(rhs_ub);
+
+    const double delta = 1.0e-11;
+    for (int i = 0; i < rhs_ub.dim(); ++i)
+    {
+        rhs_lb(i) -= delta;
+        rhs_ub(i) += delta;
+    }
+
+    // nnls.normalize_constraints(Gt, rhs_lb, rhs_ub);
+    nnls.solve_parallel_with_scalapack(Gt, rhs_lb, rhs_ub, sol);
+
+    int nnz = 0;
+    for (int i = 0; i < sol.dim(); ++i)
+    {
+        if (sol(i) != 0.0)
+        {
+            nnz++;
+        }
+    }
+
+    cout << rank << ": Number of nonzeros in NNLS solution: " << nnz
+         << ", out of " << sol.dim() << endl;
+
+    MPI_Allreduce(MPI_IN_PLACE, &nnz, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+
+    if (rank == 0)
+        cout << "Global number of nonzeros in NNLS solution: " << nnz << endl;
+
+    // Check residual of NNLS solution
+    CAROM::Vector res(Gt.numColumns(), false);
+    Gt.transposeMult(sol, res);
+
+    const double normGsol = res.norm();
+    const double normRHS = rhs_Gw.norm();
+
+    res -= rhs_Gw;
+    const double relNorm = res.norm() / std::max(normGsol, normRHS);
+    cout << rank << ": relative residual norm for NNLS solution of Gs = Gw: " <<
+         relNorm << endl;
+}
+
+// Compute EQP solution from constraints on snapshots.
+void SetupEQP_snapshots(const IntegrationRule *ir0, const int rank,
+                        ParFiniteElementSpace *fespace_X,
+                        const int nsets, const CAROM::Matrix *BV,
+                        const CAROM::Matrix *BX_snapshots, const Vector x0,
+                        const bool precondition, const double nnls_tol,
+                        const int maxNNLSnnz, NeoHookeanModel *model,
+                        CAROM::Vector &sol, CAROM::Vector *window_ids)
+{
+    const int nqe = ir0->GetNPoints();
+    const int ne = fespace_X->GetNE();
+    const int NB = BV->numColumns();
+    const int NQ = ne * nqe;
+    const int nsnap = BX_snapshots->numColumns();
+
+    MFEM_VERIFY(nsnap == BX_snapshots->numColumns() ||
+                nsnap + nsets == BX_snapshots->numColumns(), // Q: nsets?
```

Comment:
Why are there 2 possibilities?

Reply:
Actually I don't know what this is for 😅 I just copied this from the mixed nonlinear diffusion example